src/box/memtx_tx.c

```diff
@@ -1129,84 +1305,54 @@ memtx_tx_history_add_stmt(struct txn_stmt *stmt, struct tuple *old_tuple,
-	if (new_tuple != NULL) {
-		for (uint32_t i = 0; i < space->index_count; i++) {
-			struct index *index = space->index[i];
-			struct tuple **replaced = &directly_replaced[i];
-			struct tuple **successor = &direct_successor[i];
-			if (index_replace(index, NULL, new_tuple,
-					  DUP_REPLACE_OR_INSERT,
-					  replaced, successor) != 0)
-			{
-				directly_replaced_count = i;
-				goto fail;
-			}
-		}
-		directly_replaced_count = space->index_count;
-		struct tuple *replaced = directly_replaced[0];
-
-		/* Check overwritten tuple */
-		int rc;
-		if (replaced == NULL || !replaced->is_dirty)
-			rc = check_dup_clean(stmt, new_tuple, directly_replaced,
-					     &old_tuple, mode,
-					     &collected_conflicts, region);
-		else
-			rc = check_dup_dirty(stmt, new_tuple, directly_replaced,
-					     &old_tuple, mode,
-					     &collected_conflicts, region);
-		if (rc != 0)
-			goto fail;
 
-		/* Create add_story and del_story if necessary. */
-		add_story = memtx_tx_story_new_add_stmt(new_tuple, stmt);
-		if (add_story == NULL)
+	for (uint32_t i = 0; i < space->index_count; i++) {
+		struct index *index = space->index[i];
+		struct tuple **replaced = &directly_replaced[i];
+		struct tuple **successor = &direct_successor[i];
+		if (index_replace(index, NULL, new_tuple,
+				  DUP_REPLACE_OR_INSERT,
+				  replaced, successor) != 0)
+		{
+			directly_replaced_count = i;
 			goto fail;
-
-		if (replaced != NULL && !replaced->is_dirty) {
-			del_story = memtx_tx_story_new_del_stmt(replaced, stmt);
-			if (del_story == NULL)
-				goto fail;
-			del_story_created = true;
-			memtx_tx_story_link_story(add_story, del_story, 0);
-		} else if (replaced != NULL) {
-			del_story = memtx_tx_story_get(replaced);
-			memtx_tx_story_link_story(add_story, del_story, 0);
-		} else {
-			memtx_tx_handle_gap_write(stmt->txn, space,
-						  add_story, new_tuple,
-						  direct_successor[0], 0);
-		}
-
-		for (uint32_t i = 1; i < space->index_count; i++) {
-			if (directly_replaced[i] == NULL) {
-				memtx_tx_handle_gap_write(stmt->txn, space,
-							  add_story, new_tuple,
-							  direct_successor[i], i);
-				continue;
-			}
-			assert(directly_replaced[i]->is_dirty);
-			struct memtx_story *next =
-				memtx_tx_story_get(directly_replaced[i]);
-			memtx_tx_story_link_story(add_story, next, i);
-		}
-
-	} else {
-		if (old_tuple->is_dirty) {
-			del_story = memtx_tx_story_get(old_tuple);
-		} else {
-			del_story = memtx_tx_story_new_del_stmt(old_tuple,
-								stmt);
-			if (del_story == NULL)
-				goto fail;
-			del_story_created = true;
 		}
 	}
-
-	if (del_story != NULL && !del_story_created) {
-		stmt->next_in_del_list = del_story->del_stmt;
-		del_story->del_stmt = stmt;
-		stmt->del_story = del_story;
+	directly_replaced_count = space->index_count;
+	struct tuple *replaced = directly_replaced[0];
+
+	/* Check overwritten tuple */
+	int rc;
+	if (replaced == NULL || !replaced->is_dirty)
+		rc = check_dup_clean(stmt, new_tuple, directly_replaced,
```

Comment:
check_dup_clean and check_dup dirty have the same signature. Mb it is worth placing dispatching inside single check_dup()?

Reply:
The idea was to split code into parts.
Or you mean make one `check_dup` that would call either `check_dup_clean` or `check_dup_dirty`?

Reply:
Yep

Reply:
done